### DISCUSSION TITLE
Strimzi operator StrimziPodSet support

### DISCUSSION THREAD

**Discussion by roelvdberg** [argoproj/argo-cd#10096] [Q&A] [answered]
With the latest update of the strimzi operator the StrimziPodSet is enabled by default, this works similar to a statefullset and creates PVC's and pods. ArgoCD marks the PVC's for deletion and when auto pruning is on the PVC will be deleted.
https://github.com/strimzi/strimzi-kafka-operator/issues/7093

Is it possible to add the ability to ignore disks of specific applications?
I have found the option to ignore resources for a complete ArgoCD instance but that will impact other deployments. 

Another option would be to create the connection in a similar way it is done for a statefullset. This unfortunately adds knowledge about strimzi to ArgoCD so I would prefer a more generic option.

**Comment by jannfis** [comment, marked as answer]
I think the reason behind this is that the Operator copies labels (and possibly annotations) from the Kafka Operand to the PVCs it creates, including the label Argo CD uses for identifying the resources it manages. This leads to a situation where Argo CD thinks it manages the PVC, but the resource is not in Git - thus, Argo CD marks that resources as to be pruned.

You can verify this by looking at the labels of the PVCs created by the Operator. Do they have an `app.kubernetes.io/instance` label (or whatever you are currently using as the applicationLabelInstanceKey value)?

If this is the case, and you are running one of the latest stable patch releases (i.e. 2.4.6 or 2.4.7), you can try to switch the [annotation based resource tracking](https://argo-cd.readthedocs.io/en/stable/user-guide/resource_tracking/#additional-tracking-methods-via-an-annotation). With this resource tracking method, the identifier carries more meta-data about the resource, and Argo CD can make a more informed decision whether a resource should be pruned or not (e.g. the resource was created with labels copied from another resource, etc).

**Reply by scholzj** [reply]
@jannfis So - maybe a stupid question from someone who does not really use Argo - which Argo CD labels should be not copied? Do you use some Argo domain that we can simply wildcard and not copy? Or do you track it purely based on `app.kubernetes.io/instance` label?

The Kubernetes labels are quite badly specified. But IMHO copying `app.kubernetes.io/instance` label on resources belonging to some _instance_ (in this case a Kafka cluster) makes sense as the PVCs belong to that instance. But what is worse, I know of many users who rely on this for example for cost management. So I do not think we can simply stop copying it as that might fix Argo, but it would break other tools.

**Reply by jannfis** [reply]
@scholzj I agree that the choice to use `app.kubernetes.io/instance` as the tracking label by default may not have been the best one in hindsight, mainly for the label's ambiguity.

Your question is not stupid, and alas not easy to answer. Users can configure the name of the label that Argo CD uses for tracking its resources, and probably people are doing that (and it's even required in a multi-instance setup). However, when Argo CD sees that the `ownerReferences` field of a resource is populated, it will ignore this resource and assume it has been created on behalf of another controller (think, Deployment -> ReplicaSet -> Pod, where only the Deployment is managed by Argo CD and the child resources also carry the labels of the Deployment). I don't know how viable this would be for the Strimzi Operator to set the Operand as owner of the PVC (I'd assume both resources would exist in the same namespace), though.

We also introduced another tracking method using an annotation to track managed resources. This annotation carries enough metadata for Argo CD to detect whether it is set on the original resource, or if it has been copied over to another resource.

I wish we could just make that annotation based tracking method the new default, but this will a pretty invasive change for all users out there. Changing the tracking method in Argo CD triggers a re-sync of each and every resource by that Argo CD instance, removing the label and adding the annotation, and can lead to quite some load.

**Reply by scholzj** [reply]
Thanks for the explanation. Yeah, I understand that changing the default method is not something that can be done easily.

> However, when Argo CD sees that the ownerReferences field of a resource is populated, it will ignore this resource and assume it has been created on behalf of another controller (think, Deployment -> ReplicaSet -> Pod, where only the Deployment is managed by Argo CD and the child resources also carry the labels of the Deployment). I don't know how viable this would be for the Strimzi Operator to set the Operand as owner of the PVC (I'd assume both resources would exist in the same namespace), though.

We do use owner references where possible. However, the PVCs do not have them by default to mirror Kubernetes's default behavior when they are not deleted when the application is deleted. So that is why they are the only resource that shows up as most of the others are likely covered by the owner reference. Strimzi users can enable the owner reference to the PVCs as an optional thing. But that of course also brings risks the Kubernetes garbage collection deleted them etc.

**Comment by christian-schlichtherle** [comment]
For anyone looking here because ArgoCD wants to delete the PVCs, here is a quick-and-dirty workaround: Just create the PVCs with the expected specs yourself. You can copy/paste the PVCs as generated by the strimzi-cluster-operator and delete the non-relevant lines. After deployment, the strimzi-cluster-operator will put in the following labels for Zookeeper:

```
app.kubernetes.io/managed-by: strimzi-cluster-operator
app.kubernetes.io/name: zookeeper
```

... and for Kafka:

```
app.kubernetes.io/managed-by: strimzi-cluster-operator
```

ArgoCD will then try to replace these labels again with these value for Zookeeper:

```
app.kubernetes.io/managed-by: Helm
app.kubernetes.io/name: kafka
```

... and for Kafka:

```
app.kubernetes.io/managed-by: kafka
```

The strimzi-cluster-operator will change it eventually again, so the loop never ends unless you disable auto-sync in ArgoCD. Regardless, ArgoCD is not trying to delete your precious PVCs anymore.

**Comment by brat002** [comment]
https://github.com/strimzi/strimzi-kafka-operator/pull/4791/files there is an option to remove argo label from inherited resources

**Comment by marcoadasilvaa** [comment]
this option is working for me: https://github.com/strimzi/strimzi-kafka-operator/issues/7093#issuecomment-1592560236

ArgoCD version: 2.7.2
Strimzi version: 0.35.1
